proof/crefine/RISCV64/Syscall_C.thy

```diff
@@ -1234,55 +1246,500 @@ sorry (* FIXME RT: handleRecv_ccorres
+    apply (clarsimp simp: valid_sched_context'_def)
+   apply fastforce
+  apply (frule_tac n="scRefillTail sc" in h_t_valid_refill; fastforce?)
+    apply (clarsimp simp: valid_sched_context'_def)
+   apply fastforce
+  apply (frule rf_sr_ksCurSC)
+  apply (fastforce simp: csched_context_relation_def crefill_relation_def typ_heap_simps
+                         h_val_field_from_bytes' sc_ptr_to_crefill_ptr_def)
+  done
+
+crunch checkBudget
+  for cur_tcb'[wp]: cur_tcb'
+  (wp: crunch_wps simp: crunch_simps threadSet_cur)
 
 lemma checkBudget_ccorres:
-  "ccorres dc xfdc invs' UNIV []
-     checkBudget (Call checkBudget_'proc)"
-sorry (* FIXME RT: checkBudget_ccorres *)
+  "ccorres (\<lambda>rv rv'. rv = to_bool rv') ret__unsigned_long_'
+     invs' UNIV hs checkBudget (Call checkBudget_'proc)"
+  supply Collect_const[simp del]
+  apply cinit
+   apply (rule ccorres_pre_getCurSc)
+   apply (rule ccorres_pre_getConsumedTime)
+   apply (ctac (no_vcg) add: refill_sufficient_ccorres, rename_tac sufficient sufficient')
+    apply (rule_tac P=sufficient in ccorres_cases; clarsimp)
+     (* the current sc is sufficient, so we check whether the current domain has expired, and then
+        return True if it has not expired and False if it has expired *)
+     apply (rule ccorres_cond_seq)
+     apply (rule ccorres_cond_true)
+     apply (rule ccorres_rhs_assoc)+
+     apply (ctac (no_vcg) add: isCurDomainExpired_ccorres, rename_tac expired expired')
+      apply (rule_tac P=expired in ccorres_cases; clarsimp)
+       apply (rule ccorres_cond_seq)
+       apply (rule ccorres_cond_true)
+       apply ccorres_rewrite
+       apply (fastforce intro: ccorres_return_C)
+      apply ccorres_rewrite
+      apply (rule ccorres_cond_seq)
+      apply (rule ccorres_cond_false)
+      apply ccorres_rewrite
+      apply (fastforce intro: ccorres_return_C)
+     apply wpsimp
+    (* the current sc is not sufficient, so we call chargeBudget and then return False *)
+    apply (rule ccorres_cond_seq)
+    apply (rule ccorres_cond_false)
+    apply (rule ccorres_pre_getConsumedTime)
+    apply ccorres_rewrite
+    apply (ctac (no_vcg) add: chargeBudget_ccorres)
+     apply (fastforce intro: ccorres_return_C)
+    apply wpsimp+
+  apply (fastforce simp: to_bool_def)
+  done
 
 lemma checkBudgetRestart_ccorres:
-  "ccorres dc xfdc invs' UNIV []
+  "ccorres (\<lambda>rv rv'. rv = to_bool rv') ret__unsigned_long_'
+     (invs' and (\<lambda>s. weak_sch_act_wf (ksSchedulerAction s) s)) UNIV hs
      checkBudgetRestart (Call checkBudgetRestart_'proc)"
-sorry (* FIXME RT: checkBudgetRestart_ccorres *)
+  supply Collect_const[simp del]
+  apply cinit
+   apply (rule ccorres_stateAssert)
+   apply (ctac (no_vcg) add: checkBudget_ccorres, rename_tac result result')
+    apply (rule ccorres_pre_getCurThread)
+    apply csymbr
+    apply clarsimp
+    apply (rule_tac P="to_bool result'" in ccorres_cases)
+     (* checkBudget returned True, so we simply return True *)
```

Comment:
um... there is no need to say "so we" for everything that occurs; "checkBudget returned True -> return True" is sufficient

below: "so if the current thread is runnable, we set its thread state to Restart, and then return False" -> "; when the current thread is runnable, set its thread state to Restart and return False"

It's good language for papers, but the comments are describing imperative code. Pretty please consider, since it is a pattern appearing in this PR in many places.